File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        # brute force
        # max_profit = 0

        # for i in range(len(prices)-1):
        #     for j in range(i+1, len(prices)):
        #         max_profit = max(max_profit, prices[j]-prices[i])
        
        # return max_profit



        min_price = float(inf)
        max_profit = -float(inf)

        for i, price in enumerate(prices):
            min_price = min(min_price, price)
            max_profit = max(max_profit, price- min_price)
        
        return max_profit




        

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        
        
//         //brute force
//         int result = 0;
        
//         for(int i=0; i< prices.length-1; i++) {
//             for(int j=i+1; j<prices.length; j++) {
                
//                 result = Math.max(result, prices[j]-prices[i]);
//             }
//         }
//         return result;
        
        int result = 0;
        int min_price = Integer.MAX_VALUE;
        
        for(int price: prices) {
            
            min_price = Math.min(min_price, price);
            result = Math.max(result, price - min_price);
            
        }
        
        return result;
    }
}